src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -401,6 +501,162 @@ private static void AssertNoImplicitInterpolatedStringHandlerConversions(Immutab
             }
         }
 
+        private ImmutableArray<BoundExpression>? TryConvertPartsFromHandlerToConcat(ImmutableArray<BoundExpression> optimizedParts)
+        {
+
+            if (optimizedParts.Length >= 5)
+            {
+                // Prefers handler calls to concat (string[]) for 5 operands or more.
+                return null;
+            }
+            if (!optimizedParts.All(p => p is BoundCall { Method.Name: BoundInterpolatedString.AppendLiteralMethod } or BoundCall { Method.Name: BoundInterpolatedString.AppendFormattedMethod, Arguments: [{ Type.SpecialType: SpecialType.System_String }] }))
+            {
+                return null;
+
+            }
+            var builder = ArrayBuilder<BoundExpression>.GetInstance();
+            foreach (var part in optimizedParts)
+            {
+                var partCall = (BoundCall)part;
+                Debug.Assert(partCall.Method.Name != BoundInterpolatedString.AppendLiteralMethod
+                    || partCall.Arguments[0] is BoundLiteral);
+                builder.Add(
+                    partCall.Method.Name == BoundInterpolatedString.AppendLiteralMethod
+                    ? partCall.Arguments[0]
+                    : new BoundStringInsert(partCall.Syntax, partCall.Arguments[0], null, null, false)
+                );
+            }
+            return builder.ToImmutableAndFree();
+        }
+
+        private (ImmutableArray<BoundExpression> Calls, int IncreasedLiteralLength, int FilledHolesCount) OptimizeAppendFormattedCalls(ImmutableArray<BoundExpression> parts)
+        {
+            var result = ArrayBuilder<BoundExpression>.GetInstance();
+            var currentStringConst = PooledStringBuilder.GetInstance();
+            (BoundCall? AppendLiteralCall, BoundLiteral? StringLiteral) reusable = default;
+            BoundExpression? lastAppendReceiver = null;
+            int filledHolesCount = 0;
+            int increasedLiteralLength = 0;
+
+            BoundExpression createAppendLiteralCallFromBoundLiteral(BoundLiteral literal, BoundExpression receiver) =>
```

Comment:
Local functions go at the end of the containing method please. Also, please put newlines between them for readability.